pkg/azure/validation/validation.go

```diff
@@ -0,0 +1,288 @@
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/utils"
+)
+
+const providerAzure = "Azure"
+
+// ValidateMachineClassProvider checks if the Provider in MachineClass is Azure.
+// If it is not then it will return an error indicating that this provider implementation cannot fulfill the request.
+func ValidateMachineClassProvider(mcc *v1alpha1.MachineClass) error {
+	if mcc.Provider != providerAzure {
+		return status.Error(codes.InvalidArgument, fmt.Sprintf("Request for provider %s cannot be fulfilled. Only %s provider is supported.", mcc.Provider, providerAzure))
+	}
+	return nil
+}
+
+// ValidateProviderSpec validates the api.AzureProviderSpec.
+func ValidateProviderSpec(spec api.AzureProviderSpec) field.ErrorList {
+	var allErrs field.ErrorList
+	specPath := field.NewPath("providerSpec")
+
+	if len(strings.TrimSpace(spec.Location)) == 0 {
+		allErrs = append(allErrs, field.Required(specPath.Child("location"), "must provide a location"))
+	}
+	if len(strings.TrimSpace(spec.ResourceGroup)) == 0 {
+		allErrs = append(allErrs, field.Required(specPath.Child("resourceGroup"), "must provide a resourceGroup"))
+	}
+
+	allErrs = append(allErrs, validateSubnetInfo(spec.SubnetInfo, specPath.Child("subnetInfo"))...)
+	allErrs = append(allErrs, validateProperties(spec.Properties, specPath.Child("properties"))...)
+	allErrs = append(allErrs, validateTags(spec.Tags, specPath.Child("tags"))...)
+
+	return allErrs
+}
+
+// ValidateProviderSecret validates the secret containing the config to create Azure API clients.
+func ValidateProviderSecret(secret *corev1.Secret) field.ErrorList {
+	var allErrs field.ErrorList
+	secretDataPath := field.NewPath("data")
+	if utils.IsEmptyString(string(secret.Data[api.ClientID])) && utils.IsEmptyString(string(secret.Data[api.AzureClientID])) && utils.IsEmptyString(string(secret.Data[api.AzureAlternativeClientID])) {
+		allErrs = append(allErrs, field.Required(secretDataPath.Child("clientID"), "must provide clientID"))
+	}
+
+	if utils.IsEmptyString(string(secret.Data[api.ClientSecret])) && utils.IsEmptyString(string(secret.Data[api.AzureClientSecret])) && utils.IsEmptyString(string(secret.Data[api.AzureAlternativeClientSecret])) {
+		allErrs = append(allErrs, field.Required(secretDataPath.Child("clientSecret"), "must provide clientSecret"))
+	}
+
+	if utils.IsEmptyString(string(secret.Data[api.SubscriptionID])) && utils.IsEmptyString(string(secret.Data[api.AzureSubscriptionID])) && utils.IsEmptyString(string(secret.Data[api.AzureAlternativeSubscriptionID])) {
+		allErrs = append(allErrs, field.Required(secretDataPath.Child("subscriptionID"), "must provide subscriptionID"))
+	}
+
+	if utils.IsEmptyString(string(secret.Data[api.TenantID])) && utils.IsEmptyString(string(secret.Data[api.AzureTenantID])) && utils.IsEmptyString(string(secret.Data[api.AzureAlternativeTenantID])) {
+		allErrs = append(allErrs, field.Required(secretDataPath.Child("tenantID"), "must provide tenantID"))
+	}
+
```

Comment:
there is no validation for `userData` validation, we see it present in the earlier code. Could you add that.
function is `ValidateProviderSecret()`, so maybe a re-name for this function would be required. `ValidateSecret()` could be a good one.

Reply:
Ah, secrets are merged in the caller, was searching for it in mcm-azure. I have added the validation in the same method and also adjusted the tests now